lib/SPIRV/SPIRVLowerBitCastToNonStandardType.cpp

```diff
@@ -0,0 +1,200 @@
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements lowering of BitCast to nonstandard types. LLVM
+// transformations bitcast some vector types to scalar types, which are not
+// universally supported across all targets. We need ensure that "optimized"
+// LLVM IR doesn't have primitive types other than supported by the
+// SPIR target (i.e. "scalar 8/16/32/64-bit integer and 16/32/64-bit floating
+// point types, 2/3/4/8/16-element vector of scalar types").
+//
+//===----------------------------------------------------------------------===//
+#define DEBUG_TYPE "spv-lower-bitcast-to-nonstandard-type"
+
+#include "SPIRVInternal.h"
+
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/PassManager.h"
+
+#include <utility>
+
+using namespace llvm;
+
+namespace SPIRV {
+
+static VectorType *getVectorType(Type *Ty) {
+  if (!Ty)
+    return nullptr;
+  if (auto *ElemTy = dyn_cast<PointerType>(Ty))
+    Ty = ElemTy->getElementType();
+  return dyn_cast<VectorType>(Ty);
+}
+
+/// Since SPIR-V does not support non-standard vector types, instructions using
+/// these types should be replaced in a special way to avoid using of
+/// unsupported types.
+/// lowerBitCastToNonStdVec function is designed to avoid using of bitcast to
+/// unsupported vector types instructions and should be called if similar
+/// instructions have been encountered in input LLVM IR.
+bool lowerBitCastToNonStdVec(Instruction *Inst,
+                             std::vector<Instruction *> &InstsToErase) {
+  Instruction *OldInstIter(Inst);
+  IRBuilder<> Builder(OldInstIter);
+  Value *NewValue = OldInstIter->getOperand(0);
+  VectorType *NewVecTy = getVectorType(NewValue->getType());
+  InstsToErase.push_back(OldInstIter);
+  // Handle addrspacecast instruction after bitcast if present
+  for (auto *U : OldInstIter->users()) {
+    if (auto *ASCastInst = dyn_cast<AddrSpaceCastInst>(U)) {
+      unsigned DestAS = ASCastInst->getDestAddressSpace();
+      auto *NewVecPtrTy = NewVecTy->getPointerTo(DestAS);
+      // AddrSpaceCast is created explicitly instead of using method
+      // IRBuilder<>.CreateAddrSpaceCast because IRBuilder doesn't create
+      // separate instruction for constant values. Whereas SPIR-V translator
+      // doesn't like several nested instructions in one.
+      NewValue = new AddrSpaceCastInst(NewValue, NewVecPtrTy);
+      Builder.Insert(NewValue);
+      OldInstIter = ASCastInst;
```

Comment:
What if there are several addressspace casts which use the same instruction?

Reply:
Added _TODO_ cooment in [587cc7f](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1117/commits/587cc7f2a52ec020fe24677c46824f4973f58653)

Reply:
> What if there are several addressspace casts which use the same instruction?

You mean we don't need `break` at the end of the `if`? If yes, I agree.

Reply:
In this case whole function should be refactored to process several patterns "BitCast -> AScast -> Load -> ExtractElement" in an IR function.

Reply:
Yes, we can't just remove the `break`, because the code assumes that there is only one instruction which needs to be fixed